text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+  however, aim for compatibility with or supersetting of any particular C++
+  ABI.
+- Many, many interface description languages (IDLs).
+- The [x86-64 psABI](https://gitlab.com/x86-psABIs/x86-64-ABI). While we're not
+  specifying the lowering all the way to specific architectures, we can still
+  learn from how it handles various types.
+- The [aarch64 ABI](https://github.com/ARM-software/abi-aa/).
+- The ABIs of other targets.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Once we have support for extern types (such as via
+  https://github.com/rust-lang/rfcs/pull/3396), pointers and references to
+  extern types should work in crABI, just like any other opaque pointer.
+
+- Should the `improper_crabi_types` lint be a hard error, instead?
+
+- Is there a better way we can handle tuple types? Having to use a distinct
+  syntax like `cr#()` is onerous; one of the primary values of tuples is
+  brevity. In the future, if we have variadic generics, we could potentially
+  use a named tuple type, but that only helps with orthogonality, not brevity.
+  Or we could guarantee the in-memory layout of tuples *in general* in Rust. We
+  could also omit support for tuples entirely, but that seems like an
+  unfortunate gap between Rust and crABI.
+
+  We might be able to provide a shorthand, where `repr(crabi)` types and
+  `extern "crabi"` functions imply the use of crABI tuples in their type
+  signatures, but we'd still need a way of naming the distinct type, and the
+  shorthand could introduce confusion in the form of distinct types spelled
+  identically.
+
+  Would an approach based on "type transformers" help here?
+
+- The handling of `Option` and `Result` seems similarly unfortunate. We have to
+  use a distinct type to avoid freezing the layout of these Rust types for all
+  time. But could we do better, somehow? Would an approach based on "type
+  transformers" help here?
+
+- Should we provide additional guaranteed niche optimizations for `Option` in
+  crABI 1.0? In order to keep new minor revisions of crABI compatible with
+  crABI 1.0, any existing type that doesn't have a niche optimization in crABI
+  1.0 can never switch to having a niche optimization until a new
+  (incompatible) major version, and ideally we want new crABI major versions to
+  be rare.
+
+  That said, we should probably keep *most* uses of niche optimizations rare,
+  as they require special-case handling in other languages.
+
+- Similarly, should we provide additional guaranteed niche optimizations for
+  `Result`? Which ones? The same issue applies: any optimizations we want to
+  provide for `Result` on existing types should be in crABI 1.0.
+
+- Should we provide *fewer* niche optimizations? Those for `NonZero` and
+  reference types provide obvious value; are those for `bool` and `char` really
+  useful enough to justify the special case in languages that will have to
+  handle them explicitly rather than automatically?
```

Comment:
I haven't seen `Option<bool>` in many APIs, but some string APIs use `Option<char>`. Is there a way to count how popular these types are in libraries?

Reply:
Just loosely, https://github.com/search?q=%22Option%3Cbool%3E%22&type=code shows 199k results for `Optoin<bool>` but only 24.6k for `Option<char>` https://github.com/search?q=%22Option%3Cchar%3E%22&type=code